### ISSUE TITLE
Shadow being drawn over engine's canvas

### ISSUE THREAD

**Issue by Teles1** [ocornut/imgui#6290]
```
Dear ImGui 1.89.5 WIP (18946)
--------------------------------
sizeof(size_t): 4, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=199711 --Interesting that I'm using cpp 20 things on my project and even then msvc doesn't set it's compiler flags properly : )
define: _WIN32
define: _MSC_VER=1934
define: _MSVC_LANG=202002
define: IMGUI_HAS_VIEWPORT
define: IMGUI_HAS_DOCK
--------------------------------
io.BackendPlatformName: imgui_impl_win32
io.BackendRendererName: imgui_impl_dx9
io.ConfigFlags: 0x00000000
io.ConfigViewportsNoDecoration
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x00001C0E
 HasMouseCursors
 HasSetMousePos
 PlatformHasViewports
 HasMouseHoveredViewport
 RendererHasVtxOffset
 RendererHasViewports
--------------------------------
io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,64
io.DisplaySize: 1152.00,864.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 1.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

**My Issue/Question:**

I am creating a custom UI being drawn over a custom game engine that was written in DX9 back in 2000 ish and I got most of it to work as intended using my not super advanced knowledge with DX9. My issue now is that when ImGUI draws over the game it creates a a shadow of the UI on the game's canvas. How could i go about fixing this "issue" (Issue with quotes because I am sure this was intended)?

**Screenshots/Video**

![image](https://user-images.githubusercontent.com/66783472/229291750-7e2e7c5a-29c0-4147-875c-296934732cab.png)

// Here's some code anyone can copy and paste to reproduce your issue
direct9 engine init -> https://github.com/Teles1/DirectX9-Imgui-Hook/blob/master/directx9_imgui/directx9_imgui/directx9_imgui.cpp#L24
imgui init and end_scene -> https://github.com/Teles1/DirectX9-Imgui-Hook/blob/master/directx9_imgui/directx9_imgui/directx9_imgui.cpp#L24

**Comment by ocornut** [comment]
You'll need to dig in `ImGui_ImplDX9_RenderDrawData()` to find which render state we are not restoring properly, among, well, every possibly DX9 state.

**Comment by Teles1** [comment]
> You'll need to dig in `ImGui_ImplDX9_RenderDrawData()` to find which render state we are not restoring properly, among, well, every possibly DX9 state.

That is more of a feature isn't it? I don't think it's ImGui doing something wrong I just think that this is something to do with lighting? I know that because that's the same "path" that object's shadow take.. Any ideas on how to disable this behaviour before calling the original end scene?

**Comment by ocornut** [comment]
> That is more of a feature isn't it?

It is a feature that the backend should restore all its state indeed, we ought to improve/fix things if that's not perfectly the case.

> Even doing something silly like, grabbing all the states that imgui changes and saving their previous state, letting imgui render, and setting them all back to what they were doesn't solve the "issue"

Then it is something else and you'll probably need to investigate this further.
What I would do is to selectively comment/disable part of the `ImGui_ImplDX9_RenderDrawData()` function to find which call is causing this side-effect. You may use dynamic branching e.g. `if (ImGui::GetIO().KeyShift == false) { /* ignore this section is SHIFT is not pressed */`. etc.

**Comment by Teles1** [comment]
> It is a feature that the backend should restore all its state indeed, we ought to improve/fix things if that's not perfectly the case.
Oh I think I wasn't clear. I said that I don't believe that this is not ImGui issue but rather just the way that the game works. It does have lighting and it does have shadows and that makes be believe that since the UI is being drawn over the game directx9(or the game i can't tell without having source code) the UI is being "treated" as a object and shadows are being created.

With that said I was asking if you could perhaps point me on a direction to what to look for. I'm very familiar with networking, game logic etc but I definitely need to learn more regarding 3d, rendering etc.

> Then it is something else and you'll probably need to investigate this further.
What I would do is to selectively comment/disable part of the ImGui_ImplDX9_RenderDrawData() function to find which call is causing this side-effect. You may use dynamic branching e.g. if (ImGui::GetIO().KeyShift == false) { /* ignore this section is SHIFT is not pressed */. etc.

I will be trying this over the weekend and I'll get back once the issue is figured out so perhaps i can help someone in the future.

**Comment by GamingMinds-DanielC** [comment]
Sounds like something you should use RenderDoc for:
https://renderdoc.org/

A great tool to check anything related to draw calls, you can see exactly what goes wrong if you look at the render states and bound resources when comparing captures with and without ImGui.

**Comment by GamingMinds-DanielC** [comment]
> Quick edit, after checking it out it seems like RenderDoc doesn't support d3d9. No worries tho. I'll still debug it manually.

It should though, even D3D8 is still supported. Maybe it is because of how you try to use it. Hooking into a running process can be unreliable in my experience, but launching the target app from within RenderDoc (the "Launch Application" tab) should work.

Or maybe support is broken because of some reasons and nobody noticed since D3D9 is going out of style. Can't say for sure, haven't tested anything below DX11 in a while. But if that is the case, trying an older version might help. Good luck.

**Comment by Teles1** [comment]
> > Quick edit, after checking it out it seems like RenderDoc doesn't support d3d9. No worries tho. I'll still debug it manually.
> 
> It should though, even D3D8 is still supported. Maybe it is because of how you try to use it. Hooking into a running process can be unreliable in my experience, but launching the target app from within RenderDoc (the "Launch Application" tab) should work.
> 
> Or maybe support is broken because of some reasons and nobody noticed since D3D9 is going out of style. Can't say for sure, haven't tested anything below DX11 in a while. But if that is the case, trying an older version might help. Good luck.

![image](https://user-images.githubusercontent.com/66783472/230674855-9fab0e77-6343-43cf-b4da-d4f713db28fd.png)

**Comment by Teles1** [comment]
I think I'll have to take a loss here. Unfortunately I cannot figure out what is causing this behavior. I tried grabbing every single state and saving it before/after imgui call and it is identical.

**Comment by ocornut** [comment]
But have you tried what i suggested in https://github.com/ocornut/imgui/issues/6290#issuecomment-1495579677 ?

**Comment by Teles1** [comment]
> But have you tried what i suggested in [#6290 (comment)](https://github.com/ocornut/imgui/issues/6290#issuecomment-1495579677) ?

I did try it yes. 
Again, most likely not the smartest way to go about it but i added a bunch of if statements counting a variable and each number would "disable" portions of the code, my findings: Disabling the section of the ImGui_ImplDX9_SetupRenderState where it sets all the states removes a few of the features but the shadowing remains. 
// Setup render state: fixed-pipeline, alpha-blending, no face culling, no depth testing, shade mode (for gradient), bilinear sampling.
Tried:
```
if (ImGui::GetIO().KeyCtrl == true) {
    // Restore the DX9 transform
    bd->pd3dDevice->SetTransform(D3DTS_WORLD, &last_world);
    bd->pd3dDevice->SetTransform(D3DTS_VIEW, &last_view);
    bd->pd3dDevice->SetTransform(D3DTS_PROJECTION, &last_projection);
}
...
if (ImGui::GetIO().KeyCtrl == true) {
    // Restore the DX9 transform
    bd->pd3dDevice->SetTransform(D3DTS_WORLD, &last_world);
    bd->pd3dDevice->SetTransform(D3DTS_VIEW, &last_view);
    bd->pd3dDevice->SetTransform(D3DTS_PROJECTION, &last_projection);
}
```
This disables the rendering completely. So i tried one by one isolated

D3DTS_WORLD -> No visual changes 
D3DTS_VIEW -> Before activating the portion of the code that calls this it wasn't drawing anything and once I activated it once it didn't do anything after.
D3DTS_PROJECTION -> This disables ImGUI completely. (Drawing out of the screen?)

Tried preventing textures from loading at 
```
const LPDIRECT3DTEXTURE9 texture = (LPDIRECT3DTEXTURE9)pcmd->GetTexID();
bd->pd3dDevice->SetTexture(0, texture);
```
Basically everything goes "invisible" as expected besides some text that I can still see "some" texture but the shadowing is still there.
bd->pd3dDevice->SetScissorRect(&r); -> No visual changes

**Comment by ocornut** [comment]
As you edit/cut through our imgui_impl_dx9 don't worry about the imgui rendering being missing, the only thing you should care about is finding the minimum amount of DX9 calls that creates your shadowing issue.

**Comment by Teles1** [comment]
> Right, i forgot there’s a bit of a chicken and egg problem here. It looks like the framebuffer where your imgui output is rendered into is used as part of the shadow rendering, so you should still be able to swap some stuff and find hints.

I just finished another chicken/egg hunt here and I literally have no clue what the heck is going on. All i know is that this is not a bug but a feature. I know that i have to prevent directx from considering imgui's renders as objects that should have shadow. How to do that? No idea..
